New RFC: proc-macro-attribute-recursion (#2628)
rust-lang/rfcs: text/0000-proc-macro-attribute-recursion.md

- RFC PR: (leave this empty)
- Rust Issue: (leave this empty)

# Summary
[summary]: #summary

Expand `proc_macro_attributes` recursively.

# Motivation
[motivation]: #motivation

Currently, procedural macros have no way to expand macros at all. [RFC #2320](https://github.com/rust-lang/rfcs/pull/2320) aims to rectify this, but despite being reworked a lot still suffers from some complexity.

This proc_macro author wants something workable now instead of waiting for that RFC while leaving the doors open for an eventual implementation. Also making this small part available allows us to collect experience with macro expansion in proc_macros at very modest cost.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

`proc_macro_attributes` can add other attributes in their output. The expander then expands all macros in the proc_macro output, and then checks again for any attributes, expanding them, too (possibly recursively):

Here's an example from [flamer](https://crates.io/crates/flamer):

```rust
use flamer::flame;

macro_rules! macro_fun {
    () => {
        fn this_is_fun(x: u64) -> u64 {
            x + 1
        }
    }
}

#[flame] macro_fun!();
```

> Flamer expands the `#[flame] macro_fun!();` to itself (including the `#[flame]` attribute). The expander then expands `macro_fun!()` yielding

Reviewer: Wouldn't this just cause infinite expansion recursion since the compiler would again try to expand the `#[flame]` attribute?

Author: No, because flamer will just annotate `ItemMac`, not `ItemFn`. Note that the second time around, the macro has been expanded.

Author: I changed the example to make it clearer that a) the position of the `#[flame]` attribute in the expansion need not be the same as that of the attribute being expanded.

```rust
#[flame]
fn this_is_fun(x: u64) -> u64 {
    x + 1
}
```

> Because of the `#[flame]` attribtue, this is fed back to flamer, which modifies the function resulting from the macro. Note that the attribute needs not be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).

Reviewer: ```suggestion
Because of the `#[flame]` attribute, this is fed back to flamer, which modifies the function resulting from the macro. Note that the attribute does not need to be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).
```

> - Feature Name: proc_macro_attribute_recursion

Reviewer 2: ```suggestion
- Feature Name: `proc_macro_attribute_recursion`
```

> - Start Date: 24.01.2019

Reviewer 2: ```suggestion
- Start Date: 2019-01-24
```

```rust
use flamer::flame;

macro_rules! macro_fun {
    () => {
        fn this_is_fun(x: u64) -> u64 {
            x + 1
        }
    }
}

#[flame]
mod fun {
    macro_fun!();
}
```

Flamer expands the `macro_fun!();` to `#[flame] macro_fun!();`. The expander then expands `macro_fun!()` yielding

```rust
mod fun {
    #[flame]
    fn this_is_fun(x: u64) -> u64 {
        x + 1
    }
}
```

Because of the `#[flame]` attribute, this is fed back to flamer, which modifies the function resulting from the macro. Note that as in this example, the attribute needs not be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).

`proc_macro` writers can implement their macros in terms of `proc_macro_attributes` (which is a very roundabout way to deal with macros, but at least it would work at all) to gain the same benefits.

This way, `proc_macro_attribute`s can be deemed *recursive* like macros-by-example. Note that the macro recursion limit must also be observed by the `proc_macro_attribute` implementations.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> The expander is extended to search the expansion of `proc_macro` and `proc_macro_attributes` for other macro invocations. Those are then expanded until there are no more attributes or macro invocations left or the macro expansion limit is reached, whichever comes first.

Reviewer 2: This whole section (and the RFC in general) seems rather underspecified; I'd like to see examples of `proc_macro` added to the previous section at least. This also doesn't seem like a full description of behavior since `#[flame]` gets applied to macro expansions. This also doesn't say what happens if you write `#[flame(alpha, beta)]`. This should be specified including with examples.

Author: I didn't detail what `#[flame(alpha, beta)]` does (currently it should give you an error), because this RFC does not change this part of functionality. As for bang-macros, it doesn't matter if they are defined via `macro`, `macro_rules!` or as a proc macro. Again the functionality of expansion is unchanged.

The only two things this RFC defines: That the output of `proc_macro_attribute`s get expanded and that the order of expansion follows the order of appearance (so that things from the original code get expanded before things added by the expansion).

Reviewer 2: Well, from what I can tell, the reference says nothing about how `#[flame]` gets attached to `this_is_fun`.

> That the output of proc_macro_attributes get expanded and that the order of expansion follows the order of appearance (so that things from the original code get expanded before things added by the expansion).

That's not what the text says; it says "The expander is extended to search the expansion of `proc_macro` and `proc_macro_attributes` for other macro invocations." -- you've added examples for the latter but not the former.

Implementors will have to make sure to order the expansions within expanded output by their origin: macros which are in the `proc_macro_attribute`s' input need to be expanded before expanding macros that have been added by the `proc_macro_attribute`s themselves. This can easily be done by examining the `Span`s of the expansion and ordering them by `SyntaxContext`.

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer 2: The behavior of attaching `#[flame]` to expansion of macros is, as far as I can see, theoretically a breaking change if attaching the attribute has an effect on static or dynamic semantics of the expansion. I'm surprised that this behavior is not opt-in. It seems the proc macro author should request this behavior.

Author: I agree that *in theory* this is a breaking change. I' would however be surprised if anyone relied on the current behavior, as it doesn't do anything useful. As I commented on #2320, this was the thing I tried because it felt natural and just might have worked – so in effect by adding the attribute, I *am* opting in to the expansion. What else would a proc macro author expect when adding an attribute that is expanded by a proc macro?

Reviewer 2: > I' would however be surprised if anyone relied on the current behavior, as it doesn't do anything useful.

We should at least crater run it and the theoretical breakage should be added to the text with a reasoning about why you would be surprised and why it doesn't do anything useful.

> As I commented on #2320, this was the thing I tried because it felt natural and just might have worked – so in effect by adding the attribute, I _am_ opting in to the expansion. What else would a proc macro author expect when adding an attribute that is expanded by a proc macro?

It could either expand, as per your RFC, to:
```rust
mod fun {
    #[flame]
    fn this_is_fun(x: u64) -> u64 { x + 1 }
}
```
or alternatively:
```rust
mod fun {
    fn this_is_fun(x: u64) -> u64 { x + 1 } // This is what I'd expect.
}
```
this might make a world of difference if `#[flame]` does transformations to `this_is_fun` or not.

Author: OK, I added some text to this effect. Please be aware that this is a very language-lawyerly perspective (not that there's anything wrong with that) – as a proc macro author I certainly wouldn't write code to add an attribute that does nothing. In fact, I wrote this RFC because this was what I tried to get macros expanded in [flamer](https://github.com/llogiq/flamer) and I don't want to wait for a more general solution that may take far more effort.

Anyway, a crater run certainly won't hurt.
